filters/H5Zbitshuffle/src/H5Zbitshuffle.jl

```diff
@@ -0,0 +1,282 @@
+                
+                if compress_flag == BSHUF_H5_COMPRESS_LZ4
+                    err = ccall((:bshuf_compress_lz4,libbitshuffle),Cint,
+                                (Ptr{Cvoid},Ptr{Cvoid},Cuint,Cuint,Cuint),
+                                in_buf,out_buf+12,size,elem_size,block_size)
+                else
+                    err = ccall((:bshuf_compress_zstd,libbitshuffle),Cint,
+                                (Ptr{Cvoid},Ptr{Cvoid},Cuint,Cuint,Cuint),
+                                in_buf,out_buf+12,size,elem_size,block_size)
+                end
+                
+                nbytes_out = err + 12
+            end
+        else # just the shuffle thanks
+            
+            if flags & H5Z_FLAG_REVERSE != 0
+                err = ccall((:bshuf_bitunshuffle,libbitshuffle),Cint,
+                            (Ptr{Cvoid},Ptr{Cvoid},Cuint,Cuint,Cuint),
+                            in_buf,out_buf,size,elem_size,block_size)
+            else
+                err = ccall((:bshuf_bitshuffle,libbitshuffle),Cint,
+                            (Ptr{Cvoid},Ptr{Cvoid},Cuint,Cuint,Cuint),
+                            in_buf,out_buf,size,elem_size,block_size)
+            end
+            
+            nbytes_out = nbytes
+        end
+        
+        # And wrap it up
+
+        if err < 0
+            error("h5plugin_bitshuffle: Error in bitshuffle with code $err")
+        end
+        
+        Libc.free(unsafe_load(buf))
+        unsafe_store!(buf,out_buf)
+        unsafe_store!(buf_size,Csize_t(buf_size_out))
+        out_buf = C_NULL
+        
+    catch e
+
+        # On failure, return 0 and change no arguments
+        
+        nbytes_out = Csize_t(0)
+        @error "Non-fatal H5 bitshuffle plugin error: " e
+        display(stacktrace(catch_backtrace()))
+
+    finally
+        if out_buf != C_NULL
+            Libc.free(out_buf)
+        end
+    end
+
+    return Csize_t(nbytes_out)
+end
+
+# Filter registration
+
+# All information for the filter
+
+struct BitshuffleFilter <: Filter
+    major::Cuint
+    minor::Cuint
+    typesize::Cuint
+    blocksize::Cuint
+    compression::Cuint
+    comp_level::Cuint #Zstd only
+end
+
+"""
+    BitshuffleFilter(blocksize=0,compression="",comp_level=0)
+
+The Bitshuffle filter can optionally include compression "lz4" or "zstd". For "zstd"
+comp_level can be provided. This is ignored for "lz4" compression. If `blocksize`
+is zero the default bitshuffle blocksize is used.
+"""
+BitshuffleFilter(;blocksize = 0, compressor="",comp_level=0) = begin
+    lowercase(compressor) in ("lz4","zstd","") || throw(ArgumentError("Invalid bitshuffle compression $compression"))
+    compcode = 0
+    if compressor == "lz4" compcode = BSHUF_H5_COMPRESS_LZ4
+    elseif compressor == "zstd" compcode = BSHUF_H5_COMPRESS_ZSTD
+    end
+    BitshuffleFilter(BSHUF_VERSION_MAJOR,BSHUF_VERSION_MINOR,0,blocksize,compcode,comp_level)
+end
```

Comment:
The formatting here is strange. Could we just use the normal `function` syntax here?

```suggestion
function BitshuffleFilter(;blocksize = 0, compressor="",comp_level=0)
    lowercase(compressor) in ("lz4","zstd","") || throw(ArgumentError("Invalid bitshuffle compression $compression"))
    compcode = 0
    if compressor == "lz4"
        compcode = BSHUF_H5_COMPRESS_LZ4
    elseif compressor == "zstd"
        compcode = BSHUF_H5_COMPRESS_ZSTD
    end
    BitshuffleFilter(BSHUF_VERSION_MAJOR,BSHUF_VERSION_MINOR,0,blocksize,compcode,comp_level)
end
```

Reply:
```suggestion
function BitshuffleFilter(;blocksize = 0, compressor=:none, comp_level=0)
    if isa(compressor, String)
       compressor = Symbol(lowercase(compressor))
    end
    compressor in (:lz4, :zstd, :none) || throw(ArgumentError("Invalid bitshuffle compression $compression"))
    compcode = 0
    if compressor == :lz4
        compcode = BSHUF_H5_COMPRESS_LZ4
    elseif compressor == :zstd
        compcode = BSHUF_H5_COMPRESS_ZSTD
    end
    BitshuffleFilter(BSHUF_VERSION_MAJOR,BSHUF_VERSION_MINOR,0,blocksize,compcode,comp_level)
end```